arangod/Transaction/Methods.cpp

```diff
@@ -1697,13 +1704,16 @@ OperationResult transaction::Methods::insertLocal(
       }
 
       replicationType = ReplicationType::LEADER;
+      if (isMMFiles && needsLock) {
+        keyLockInfo.shouldLock = true;
+      }
       // We cannot be silent if we may have to replicate later.
       // If we need to get the followers under the single document operation's
       // lock, we don't know yet if we will have followers later and thus cannot
       // be silent.
       // Otherwise, if we already know the followers to replicate to, we can
       // just check if they're empty.
-      if (needsToGetFollowersUnderLock || !followers->empty()) {
+      if (needsToGetFollowersUnderLock || keyLockInfo.shouldLock || !followers->empty()) {
```

Comment:
Similar to the above, `needsToGetFollowersUnderLock` should be equal to `keyLockInfo.shouldLock`. You could add an assert for that and/or omit the additional condition.

Reply:
Even if these things are actually related, I don't want this coupling here. It does no harm to explicitly state the extra condition here, and IMHO it makes it clearer than having a single variable with multiple responsibilities.

Reply:
I'm probably just mixing those conceptually because both flags (`needsToGetFollowersUnderLock` and `keyLockInfo.shouldLock`) solve problems arising from no ("normal") lock being held during replication. And as above, I just wanted to note this and I don't have a strong opinion either way.